### ISSUE TITLE
Wrong URL when opening file on Sourcegraph web using the JetBrains plugin

### ISSUE THREAD

**Issue by taras-yemets** [sourcegraph/sourcegraph#49510]
- **Platform information:**  Sourcegraph for JetBrains `2.1.2`

#### Steps to reproduce:

1. Open the file from the Perforce repo in JetBrains IDE.
2. Ensure Sourcegraph plugin is installed and configured ([README](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1295ad99e2c682e8374648f2cc5c18eb3a393d6b/-/blob/client/jetbrains/README.md)).
3. Open Sourcegraph plugin context menu and click "Open selection in Sourcegraph web".

More context can be found in [this Slack thread](https://sourcegraph.slack.com/archives/C01P4R0RZH9/p1670909327330639). 

#### Expected behavior:
The file is opened on a connected Sourcegraph instance with proper query params (revision, line, etc.).

#### Actual behavior:
The link navigates to the file not found page on a connected Sourcegraph instance.

If you would like immediate help on this, please email support@sourcegraph.com (you can still create the issue, but there are no [SLAs](https://about.sourcegraph.com/support/) on issues like there are for support requests).

**Comment by peterguy** [comment]
Does this have any overlap with #42294?

**Comment by vdavid** [comment]
- Added the customer-specific infos here: https://github.com/sourcegraph/customer/issues/2146

**Comment by vdavid** [comment]
@peterguy: they have no overlap.
- https://github.com/sourcegraph/sourcegraph/issues/42294, is about the situation when one uses the "Find with Sourcegraph" dialog (the one that opens for `Option+S`), finds a file, then chooses to open the file for editing. In this case, we try to match the file found on Sourcegraph (which only exists in memory) to a _local_ file (which exists on the disk). The benefit of doing this successfully is, of course, that you can make changes to the file and eventually commit it. This is what doesn't always work with Perforce now.
- This current issue is for the case when you have a _local_ file open in your IDE, and you want to open this file on _Sourcegraph web_, typically for sharing the link with a colleague. This is another thing that doesn't work well with some Perforce setups.

**Comment by taras-yemets** [comment]
@vdavid, @philipp-spiess, and I looked into this issue. Here are some findings.

The repo on the user’s system is coming from a GHE code host (is a Git repository), but that code is cloned over from Perforce using some proprietary tooling. However, Sourcegraph instance only has Perforce code cloned via [p4-fusion](https://github.com/salesforce/p4-fusion) ([original message](https://sourcegraph.slack.com/archives/C01P4R0RZH9/p1684322250981679?thread_ts=1670909327.330639&cid=C01P4R0RZH9)).
Thus the version control system defined by [`getRemoteRepoURL`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/client/jetbrains/src/main/java/com/sourcegraph/vcs/RepoUtil.java?L81:5-98:6) is `VCSType.GIT` and is [handled respectively](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/client/jetbrains/src/main/java/com/sourcegraph/vcs/GitUtil.java?L20:1-30:6). This remote repo URL is used to [build an editor file URL](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/client/jetbrains/src/main/java/com/sourcegraph/website/URLBuilder.java?L17:1-27:6). It is later used redirected according to the logic defined in [`editor.go`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/cmd/frontend/internal/app/editor.go?L162). 
As the Sourcegraph instance knows nothing about this Git repo (it knows only about the Perforce one the Git repo has been cloned from), the user is expectedly navigated to the repo not found page (Git repo doesn't exist on the Sourcegraph instance).

There is a possibility to [configure remote URL replacements](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/client/jetbrains/README.md?L64-80&view=code). Still, it's not going to work with the current logic of opening/copying Sourcegraph URL from IDE ([building the editor URL](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/client/jetbrains/src/main/java/com/sourcegraph/website/URLBuilder.java?L17:1-27:6) in the IDE plugin and then [opening file redirect](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4a7033d667c6f0d5444b0b8dc8347cee883dc046/-/blob/cmd/frontend/internal/app/editor.go?L162)) as [`RepoSourceCloneURLToRepoName`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@58da6780d767b9e667e884356894a956bed21c9a/-/blob/internal/cloneurls/clone_urls.go?L42:1-56:3) function currently doesn't support Perforce code host.

Possible solutions we could think of:
1. Add another plugin configuration field to the JetBrains plugin defining repo name configurations. If the project opened in IDE matches the pattern from this new field value, we build the actual link to a file on Sourcegraph instance (e.g. similar to [`toBlobAbsoluteURL`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/client/shared/src/util/url.ts?L420:1-456:2)) on the plugin side (but not the [editor file URL](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/client/jetbrains/src/main/java/com/sourcegraph/website/URLBuilder.java?L16-45)). In this way we can skip the mentioned [open file redirect](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/cmd/frontend/internal/app/editor.go?L159) logic which currently doesn't support Perforce. Keep the repo remote URL replacements logic as it is.
2. Do not add a new config field and use the existing logic for repo remote URL replacements. Build the actual link to a file on Sourcegraph instance (e.g. similar to [`toBlobAbsoluteURL`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/client/shared/src/util/url.ts?L420:1-456:2)) on the plugin side. In this case, we do not depend on [`openFileRedirect`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/cmd/frontend/internal/app/editor.go?L159) logic.
3. Do not change the plugin code, but add Perforce support to [`RepoSourceCloneURLToRepoName`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f7411c1012e0b5a237591ca3cacf1708e43dd5d9/-/blob/internal/cloneurls/clone_urls.go?L26:6).

**Comment by taylorsperry** [comment]
@taras-yemets Do you know if this is still an issue? Sounds like it's related to the "search" component of the JB Search + Cody plugin.

**Comment by taras-yemets** [comment]
> @taras-yemets Do you know if this is still an issue? Sounds like it's related to the "search" component of the JB Search + Cody plugin.

@taylorsperry, yes, this is JB Search + Cody plugin and it's [still relevant](https://sourcegraph.slack.com/archives/C01P4R0RZH9/p1704771738406169?thread_ts=1670909327.330639&cid=C01P4R0RZH9).
